### PR TITLE
Fix avatars in federated calls

### PR DESCRIPTION
**Pull request by danxuliu**
Follow up to #12668

Although the `userId` property in the `participants->update` signaling message, which is used by clients to know the participants in a call, [was set to the cloud ID for federated participants](https://github.com/nextcloud/spreed/pull/12668/commits/f91bdfa38ae5922fac15bd64353715bceac8d08a#diff-7ed631e9f04a0445d27e26f716dee988bd363d56cbfa4453ed74bd3605433f9aR433-R434) this was not enough to identify a federated user, as it was not possible to know for sure the actor type only based on the user ID (as [a local user could have a user ID that resembled a cloud ID](https://github.com/nextcloud/spreed/pull/12668#discussion_r1690234134), so looking for _@ remote_ would not be reliable).

Due to that now the `participants->update` signaling messages explicitly provide the `actorType` and `actorId` of each participant in the message (and `userId` is again provided only for regular users like it was always done). Knowing the actor type and id clients can now request the avatar from the right endpoint and with the right data.

In the WebUI now `actorType` and `actorId` take precedence over `userId` when identifying a participant in `VideoView`, which automatically causes `AvatarWrapper` to use the right endpoint (once the type and id are converted, see below).

## Converting actor type and id

Conversion of `actorType` and `actorId` in `participants->update` signaling messages will be automatically done by the external signaling server, so clients just need to use the received values as is (so the client of a federated user will see `actorType: users` for users in its own server and `actorType: federated_users` for users in the host server, in both cases with `actorId` adjusted as needed, just like when getting the participants from the API). However, this might become relevant again once/if federated calls are implemented in the internal signaling server, so leaving the section below for reference.

<details>
There is an important detail to keep in mind, though. The actor type and id in the signaling messages are from the point of view of the host Nextcloud server. Therefore, the client of a federated user would need to convert the actor type and id between local users and federated users, similarly to [how a federated Nextcloud server does it](https://github.com/nextcloud/spreed/blob/897c069d4ae80a5500e6dd0ac735bf3ebbc9e33a/lib/Federation/Proxy/TalkV1/UserConverter.php#L38-L53) when returning the list of participants (as a local user from the point of view of the host is a federated user from the point of view of the federated server, and a federated user from a specific remote server from the point of view of the host is a local user from the point of view of that specific remote server).

Note, however, that in order to get the avatars only converting from local users to federated users is strictly needed. Avatars can be got using two different API endpoints, one that accepts a user ID and another one that accepts a cloud ID; in the first case only avatars for local users are returned, while in the second case avatars are returned for both local and remote users. Therefore, if a federated user is not converted to a local user the proxy endpoint will be used, but [it will correctly return the avatar for the local user](https://github.com/nextcloud/spreed/blob/1ccbf67da231c37801c1406b10000144125aec35/lib/Controller/AvatarController.php#L247-L255). Nevertheless it might be better to do the conversion anyway for any other future use of the actor type and id.
</details>

## How to test

- Set up two Nextcloud servers with external signaling servers and federation
  - External signaling servers require support for converting the `actorType` and `actorId` in `participants->update` signaling messages
- Create a conversation in Nextcloud server A
- Add two users from Nextcloud server B
- In a private window, log in Nextcloud server B with one user
- Accept the invitation and join the conversation
- In another browser/device, log in Nextcloud server B with the other user
- Accept the invitation and join the conversation
- Start the call and join with all users

### Result with this pull request

The right avatars are shown for all users

### Result without this pull request

Broken avatars are shown for all users (except oneself)

### CODE REVIEW

File: src/utils/signaling.js [hunk 1]
```diff
@@ -1499,6 +1500,26 @@ Signaling.Standalone.prototype.processRoomParticipantsEvent = function(data) {
...
+				for (let i = 0; i < data.event.update.users.length; i++) {
+					const user = data.event.update.users[i]
+					if (user.actorType === ATTENDEE.ACTOR_TYPE.USERS) {
+						user.actorType = ATTENDEE.ACTOR_TYPE.FEDERATED_USERS
+						user.actorId = user.actorId + '@' + this.settings.federation.nextcloudServer
+					} else if (user.actorType === ATTENDEE.ACTOR_TYPE.FEDERATED_USERS
+							&& (user.actorId.endsWith('@' + getBaseUrl())
+								|| user.actorId.endsWith('@' + getBaseUrl().replace('https://', ''))
+							)) {
+						let indexOfDomain = user.actorId.indexOf('@' + getBaseUrl())
```
Comment (nickvergessen) [inline src/utils/signaling.js:1514]: is there something like `last index of` 
Unlikely but could happen :see_no_evil:

File: src/utils/webrtc/webrtc.js [hunk 2]
```diff
@@ -293,6 +293,7 @@ function usersChanged(signaling, newUsers, disconnectedSessionIds) {
 				webRtc: webrtc,
 			})
 		}
+		callParticipantModel.setActor(user.actorType, user.actorId)
```
Comment (nickvergessen) [inline src/utils/webrtc/webrtc.js:296]: desktop client goes :boom: when speaking to a 19 server?

File: src/utils/signaling.js [hunk 3]
```diff
@@ -1499,6 +1500,26 @@ Signaling.Standalone.prototype.processRoomParticipantsEvent = function(data) {
 				console.error('Unknown room participant event', data)
 			}
 		} else {
+			// Convert participant data in federated conversations
+			if (this.settings.federation?.nextcloudServer) {
+				for (let i = 0; i < data.event.update.users.length; i++) {
+					const user = data.event.update.users[i]
+					if (user.actorType === ATTENDEE.ACTOR_TYPE.USERS) {
```
Comment (nickvergessen) [inline src/utils/signaling.js:1507]: Also here. Must check for existance of the attributes so the desktop client doesn't break against 19

File: src/utils/signaling.js (hunk 1 above)
Comment (danxuliu) [reply src/utils/signaling.js]: Correct, but no longer needed as commit was dropped.

File: src/utils/webrtc/webrtc.js (hunk 2 above)
Comment (danxuliu) [reply src/utils/webrtc/webrtc.js:296]: The attributes would be simply `undefined`, which would not change the previous behaviour nor break anything, no?

File: src/utils/signaling.js (hunk 3 above)
Comment (danxuliu) [reply src/utils/signaling.js]: Commit dropped.

**Comment by danxuliu** [comment]
The conversion between `actorType`and `actorId` was dropped, as @fancycode will be doing that in the signaling server. Thanks a lot!

Therefore while avatars are still broken in the current state this pull request prepares the WebUI to correctly shown them once the signaling server is updated :-)

File: src/utils/webrtc/webrtc.js (hunk 2 above)
Comment (nickvergessen) [reply src/utils/webrtc/webrtc.js:296]: ah, no PHP style `Undefined key X` error. fine by me then

**Review by nickvergessen — APPROVED** [review]

**Comment by fancycode** [comment]
> The conversion between `actorType`and `actorId` was dropped, as @fancycode will be doing that in the signaling server. Thanks a lot!

Done in https://github.com/strukturag/nextcloud-spreed-signaling/pull/776/commits/1a823efc27bae2d664a401b624a12ef7b73693b9

**Pull request merged**